server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -552,7 +552,7 @@ public void writeTo(StreamOutput out) throws IOException {
             if (out.getVersion().onOrAfter(VERSION_STORE_VALUES_AS_BYTES_REFERENCE)) {
                 out.writeBytesReference(valueRef);
             } else {
-                valueRef.writeTo(out);
+                out.writeGenericValue(new ArrayList(this));
```

Comment:
Don't we also have a problem in the constructor that takes a StreamInput as argument? Usually to test this we have a test that deserializes a binary base64 written with the previous version. Then you are really sure that the current code can decode what comes from the codebase of the previous version.

Reply:
I am not sure I follow. What problem do you see in the constructor that takes a StreamInput as argument ?

Reply:
I do not see an if based on the version, and I think it always reads a bytes ref? Or maybe I missed something?

Reply:
oh well I may have gotten lost in this very long class and looked in the wrong place :)

Reply:
I did not follow why BinaryValues uses writeGenericValue while its constructor doesn't

Reply:
The valueRef is serialized from a list by `writeGenericValue`, and the way to iterate `BinaryValues` is similar to `readGenericValue`. So it seems that `out.writeGenericValue(new ArrayList(this))` just deserialize the valueRef and serialize it again. I wonder which case can be resolved in this way ?